feat(scripts): evaluate script blocks

wmalik opened the pull request:
# Reason for This Change

The attributes of a script block need to be evaluated at runtime in order to resolve any variables/globals/functions etc.

## Description of Changes

* Added an `EvalScript` function that evaluates a script based on the provided context
* Added tests for validating types + evaluating functions/globals

i4ki commented on config/script.go:
```diff
@@ -0,0 +1,192 @@
+// care about which command attr was set e.g. command or commands. This method
+// returns a list of commands irrespective of whether they were set through
+// job.command or job.commands
+func (es EvaluatedScriptJob) CommandOrCommands() [][]string {
```
Just call it `Commands()`. The evaluated data structure does not need to reflect the source code data structures. In practice, only the list of commands make sense.

i4ki commented on config/script.go:
```diff
@@ -0,0 +1,192 @@
+	errs := errors.L()
+
+	desc, err := evalScriptDesc(evalctx, script.Description.Expr, "script.description")
+	if err != nil {
+		errs.Append(err)
+	}
+
```
```suggestion
	errs.Append(err)
```

i4ki commented on config/script.go:
```diff
@@ -0,0 +1,192 @@
+	}
+	errs := errors.L()
+
+	desc, err := evalScriptDesc(evalctx, script.Description.Expr, "script.description")
```
we have `evalString()` in this same package.
https://github.com/terramate-io/terramate/blob/main/config/assert.go#L74-L86

wmalik replied:
Yes, I considered using it however I'd like the eval* functions to return specific errors (e.g. not a general error e.g.`ErrSchema`) that are relevant to the attributes e.g. description

i4ki replied:
you can wrap the errors with `errs.AppendWrap()`, or `err.Append(errors.E(OtherErr, err))`.

wmalik replied:
I see, I'll use the `evalString()` function then. Would it be possible to print only the top level error when we later emit this to the user's stdout? 🤔

i4ki replied:
I think the wrapping makes sense (as you did now).

```
terramate schema error: invalid type for script.description
```

Do you think it should be still improved?

wmalik replied:
Ah sorry, I forgot to embed the underlying error, fixed now.
![image](https://github.com/terramate-io/terramate/assets/92498/b671b455-4c58-4c84-9611-4c93a4bedca0)

i4ki commented on config/script.go:
```diff
@@ -0,0 +1,192 @@
+
+	if !val.Type().IsTupleType() {
+		return nil, errors.E(ErrScriptInvalidTypeCommands, expr.Range(),
+			"%s should be a list type", name)
```
```suggestion
			"%s should be a list of list(string) type", name)
```

i4ki commented on config/script.go:
```diff
@@ -0,0 +1,192 @@
+			_, elem := itCommand.Element()
+			if elem.Type() != cty.String {
+				errs.Append(errors.E(ErrScriptInvalidTypeCommands, expr.Range(),
+					"field %s must be a list(string) but element %d has type %q",
```
```suggestion
					"field %s must be a string but element %d has type %q",
```

i4ki commented on config/script_test.go:
```diff
@@ -0,0 +1,308 @@
+			got, err := config.EvalScript(hclctx, tcase.script)
+			assert.IsError(t, err, tcase.wantErr)
+
+			if tcase.want.Labels != nil {
+				assert.IsTrue(t, slices.Equal(tcase.want.Labels, got.Labels),
+					fmt.Sprintf("label mismatch: want[%#v], got [%#v]", tcase.want.Labels, got.Labels))
+			}
+			assert.EqualStrings(t, tcase.want.Description, got.Description)
+			assert.EqualInts(t, len(tcase.want.Jobs), len(got.Jobs), "len(jobs) mismatch")
+
+			for i, gotJob := range got.Jobs {
+				wantJob := tcase.want.Jobs[i]
+
+				if wantJob.Command != nil {
+					assert.IsTrue(t,
+						slices.Equal(wantJob.Command, gotJob.Command),
+						fmt.Sprintf("command value mismatch: want[%#v], got [%#v]", wantJob.Command, gotJob.Command))
+				}
+
+				if wantJob.Commands != nil {
+					assert.EqualInts(t, len(wantJob.Commands), len(gotJob.Commands), "len(commands) mismatch")
+					for j, gotCommand := range gotJob.Commands {
+						wantCommand := wantJob.Commands[j]
+						assert.IsTrue(t,
+							slices.Equal(wantCommand, gotCommand),
+							fmt.Sprintf("command value mismatch: want[%#v], got [%#v]", wantCommand, gotCommand))
+					}
+				}
+
+			}
```
As `EvaluatedScript` and `EvaluatedScriptJob` are simple types (structs with basic Go types), what do you think of just using `cmp.Diff()` (go-cmp):

```
if diff := cmp.Diff(expectedScript, gotScript); diff != "" {
    t.Fatal(diff)
}
```

wmalik replied:
Good idea, that would save us some lines of code and also the diff output is much more easier to read.

i4ki reviewed (commented):
It looks very good!!

I left some comments for discussion.

snakster commented on config/script.go:
```diff
@@ -0,0 +1,192 @@
+}
+
+// EvaluatedScript represents an evaluated script block
+type EvaluatedScript struct {
```
I think the naming should align with similar types of this package, i.e.
```
// Stack represents an evaluated stack.
type Stack struct { ...

// Assert represents evaluated assert block configuration.
type Assert struct { ...
```

i4ki replied:
Very good point. I missed that.

wmalik commented:
@snakster @i4ki thanks for the review. I have addressed your feedback in these commits:

* https://github.com/terramate-io/terramate/pull/1255/commits/19193918a24ed71a2cc01ec3932942b574fe4930
* https://github.com/terramate-io/terramate/pull/1255/commits/17a7bbba647426a9580a745c5cc31de2f38c1155

Let me know if you are fine with the changes, then I will squash+push and ask for an approval.

snakster commented:
Looks good 👍 

One request, could you add the `hcl.Range` to the `config.Script` type, similar to how it's done in `config.Assert`?

It would be convenient for printing out something like `location: /terramate.tm:200,20` when listing the scripts.

i4ki reviewed (approved):
LGTM

i4ki reviewed (approved):
LGTM 🎉
